File: statisticslib/stats/KolmogorovSmirnovStatistics.py
import numpy as np
import scipy
from statisticslib.timeseries.TimeSeries import TimeSeries
from statisticslib.stats.Statistics import Statistics
from statisticslib.models.DynamicsModel import DynamicsModel


class KolmogorovSmirnovStatistics(Statistics):

    def __init__(
        self,
        testedModel: DynamicsModel,
        timeStep: int,
        backtestHorizon: int
    ):
        super().__init__(testedModel, timeStep, backtestHorizon)
        self.__eigenvalues, self.__eigenvectors = np.linalg.eig(
            self._testedModel.getReturnCorrelation(
                self._backtestHorizon, self._timeStep
            )
        )

    def getValue(
        self,
        observations: TimeSeries,
        isModified: bool = True
    ) -> float:
        """
        ### Args:
            `observations` (TimeSeries) - trajectory of process \n
            `isModified` (bool) - flag for usage modified statistic \n
        """
        # Модификация
        if isModified:
            zSeries: np.ndarray = self._testedModel.transformToStationary(
                observations, self._backtestHorizon
            ).to_numpy()

            eigenvalues: np.ndarray = 1.0 / np.sqrt(self.__eigenvalues)

            zDecorrelaredSeries: np.ndarray = (
                zSeries @ self.__eigenvectors) * eigenvalues

            uSeries: np.ndarray = scipy.stats.norm.cdf(zDecorrelaredSeries)
        else:
            uSeries: np.ndarray = self._testedModel.transformToUniform(
                observations, self._backtestHorizon
            ).to_numpy()

        sizeSeries: int = uSeries.shape[0]

        uSeriesSorted: np.ndarray = np.sort(uSeries)
        sequenceIntNum: np.ndarray = np.arange(1, sizeSeries + 1)

        # ???? без понятия как переименовать
        d_plus: float = np.max((sequenceIntNum / sizeSeries) - uSeriesSorted)
        d_minus: float = np.max(
            uSeriesSorted - ((sequenceIntNum - 1) / sizeSeries)
        )

        return np.max([d_plus, d_minus])

File: statisticslib/stats/KolmogorovSmirnovStatistics.cpp
#include "KolmogorovSmirnovStatistics.h"
#include <algorithm>
#include <cmath>

KolmogorovSmirnovStatistics::KolmogorovSmirnovStatistics(DynamicsModel& testedModel, int timeStep, int backtestHorizon)
    : Statistics(testedModel, timeStep, backtestHorizon), _testedModel(testedModel), _backtestHorizon(backtestHorizon) {}

double KolmogorovSmirnovStatistics::getValue(const std::vector<double>& observations, bool isModified) {
    std::vector<double> uSeries;

    if (isModified) {
        uSeries = _testedModel.transformToUniform(observations, _backtestHorizon);
    } else {
        uSeries = observations;
    }

    std::sort(uSeries.begin(), uSeries.end());

    int n = uSeries.size();
    double D = 0.0;

    for (int i = 1; i <= n; ++i) {
        double empirical = static_cast<double>(i) / n;
        double theoretical = uSeries[i - 1];
        double diff = std::abs(empirical - theoretical);
        if (diff > D) {
            D = diff;
        }
    }

    double theoretical_start = 0.0;
    double empirical_start = 0.0;
    double diff_start = std::abs(empirical_start - theoretical_start);
    if (diff_start > D) {
        D = diff_start;
    }

    double theoretical_end = 1.0;
    double empirical_end = 1.0;
    double diff_end = std::abs(empirical_end - theoretical_end);
    if (diff_end > D) {
        D = diff_end;
    }

    return D;
}